tests/frame_manager_test.go

```diff
@@ -33,37 +35,79 @@ func TestWaitForFrameNavigationWithinDocument(t *testing.T) {
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
-			tb := newTestBrowser(t, withFileServer())
-			p := tb.NewPage(nil)
+			t.Parallel()
 
-			resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
-			require.NotNil(t, resp)
+			errc := make(chan error, 1)
+			go func() {
+				tb := newTestBrowser(t, withFileServer())
+				p := tb.NewPage(nil)
 
-			// A click right away could possibly trigger navigation before we
-			// had a chance to call WaitForNavigation below, so give it some
-			// time to simulate the JS overhead, waiting for XHR response, etc.
-			time.AfterFunc(timeout*time.Millisecond, func() { //nolint:durationcheck
-				_ = tb.await(func() error {
-					_ = p.Click(tc.selector, nil)
-					return nil
-				})
-			})
+				resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
+				require.NotNil(t, resp)
 
-			done := make(chan struct{}, 1)
-			go func() {
-				require.NotPanics(t, func() {
-					p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
-						Timeout: timeout * 3, // interpreted as ms
+				// A click right away could possibly trigger navigation before we
+				// had a chance to call WaitForNavigation below, so give it some
+				// time to simulate the JS overhead, waiting for XHR
+				// response, etc.
+				<-time.After(timeout * time.Millisecond) //nolint:durationcheck
+
+				// if one of the promises panics, err will contain the error
+				errc <- tb.await(func() error {
+					_ = p.Click(tc.selector, nil)
+					_ = p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
+						Timeout: 3 * timeout, // interpreted as ms
 					}))
+					return nil
```

Comment:
I think we should be checking all errors to ensure that we're not seeing them -- so i guess we should be ensuring that resolve is called and not reject?

Reply:
`tb.wait()` is required for the call to `Click` and  `WaitForNavigation` to work on the event loop. The function in `tb.wait()` also needs to return before the event loop either resolves or rejects the result of `Click` and  `WaitForNavigation`. The main change i've done is to add a check to ensure that both `Click` and  `WaitForNavigation` end up in the correct fulfilled state, and this check can only be performed outside of `tb.wait()`.

Reply:
OK, that makes sense. I played around with the test, and it looks great, besides the minor suggestions I made just now.

Do you want to leave the functional enhancement (checking for an existing element post-navigation) for later? We can create an issue for it if you prefer not doing it here.

Reply:
I'm not sure how to test that something has changed when an anchor link is clicked. I was hoping that `isVisible` would do the trick, but that doesn't take into account whether the element is in the viewport or not. Any ideas what I could check for?

Also, the script that is linked to `nav-history`'s `click` event doesn't work, so i think it's best to just remove the `nav-history` test.

Reply:
> Any ideas what I could check for?

It should be straightforward to determine. E.g. [see here](https://www.javascripttutorial.net/dom/css/check-if-an-element-is-visible-in-the-viewport/). We already have `ElementHandle.boundingBox()`, so we might be able to check that in Go. Having `ElementHandle.isInViewport()` seems like a useful user-facing function to have anyway, even if Playwright doesn't.

But this is a bit out of scope for this PR, so I can create an issue if you agree, and we can work on it later.


> Also, the script that is linked to `nav-history`'s `click` event doesn't work, so i think it's best to just remove the `nav-history` test.

WDYM by "doesn't work"? If you mean that functionally nothing happens on the page, that's expected, since `history.pushState()` simply creates a new history entry in the browser, and it's up to the JS app to interpret that, which is not part of the test.

The only thing we care for is if the `Page.navigatedWithinDocument` CDP event is triggered and if we process it correctly. If you run the test with `XK6_BROWSER_LOG=trace` you can see that happening, so I think the test is still useful.

Though if we want to expand the test to functionally check if an element is within the viewport, the JS will be slightly more complicated.

Reply:
Nice, I was thinking about working with the boundingbox values on the go side, but wasn't sure if there was an easier way. Let's create a new issue for this, maybe it could coincide with working on `isInViewport` which would have helped here.

> WDYM by "doesn't work"?

Yeah, it didn't exhibit any visible behaviour. I've double checked it with `XK6_BROWSER_LOG=trace` and the `Page.navigatedWithinDocument` CDP event is triggered 👍

Reply:
Created this issue to track the functionality check part: https://github.com/grafana/xk6-browser/issues/492

Reply:
Ah, thanks. I created #493, but was debating whether there should be a separate issue to change the test, but it makes sense to have one.